hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```
@@ -3972,11 +3973,91 @@ func (r *HostedClusterReconciler) validateHostedClusterSupport(hc *hyperv1.Hoste
+// it will set the defaultIPv4 or defaultIPv6 advertise address depending on the
+// ClusterNetwork Primary IP.
+func findAdvertiseAddress(hc *hyperv1.HostedCluster) string {
+	var advertiseAddress *string
+	if hc.Spec.Networking.APIServer != nil && hc.Spec.Networking.APIServer.AdvertiseAddress != nil {
+		advertiseAddress = hc.Spec.Networking.APIServer.AdvertiseAddress
+
+		ipaddr := net.ParseIP(*advertiseAddress)
+		if ipaddr != nil {
+			return *advertiseAddress
+		}
+	}
+
+	ipaddr := net.ParseIP(hc.Spec.Networking.ClusterNetwork[0].CIDR.IP.String())
+
+	if ipaddr != nil && strings.Contains(hc.Spec.Networking.ClusterNetwork[0].CIDR.IP.String(), ".") {
+		advertiseAddress = k8sutilspointer.String(config.DefaultAdvertiseIPv4Address)
+	}
+
+	if ipaddr != nil && strings.Contains(hc.Spec.Networking.ClusterNetwork[0].CIDR.IP.String(), ":") {
+		advertiseAddress = k8sutilspointer.String(config.DefaultAdvertiseIPv6Address)
+	}
+
+	return *advertiseAddress
+}
+
+// checkAdvertiseAddressOverlapping validates that the AdvertiseAddress defined does not overlap with
+// the ClusterNetwork, ServiceNetwork and MachineNetwork
+func checkAdvertiseAddressOverlapping(hc *hyperv1.HostedCluster) field.ErrorList {
+	var errs field.ErrorList
+	var err error
+	var advAddress netip.Addr
+
+	advAddrPath := "spec.networking.APIServerNetworking.AdvertiseAddress"
+	networks := make(map[string]string, 0)
+
+	if len(hc.Spec.Networking.ClusterNetwork) > 0 {
+		networks["spec.networking.ClusterNetwork"] = hc.Spec.Networking.ClusterNetwork[0].CIDR.String()
+	}
+
+	if len(hc.Spec.Networking.ServiceNetwork) > 0 {
+		networks["spec.networking.ServiceNetwork"] = hc.Spec.Networking.ServiceNetwork[0].CIDR.String()
+	}
+
+	if len(hc.Spec.Networking.MachineNetwork) > 0 {
+		networks["spec.networking.MachineNetwork"] = hc.Spec.Networking.MachineNetwork[0].CIDR.String()
+	}
+
+	advAddress, err = netip.ParseAddr(findAdvertiseAddress(hc))
```

why are we calling netip.ParseAddr( over findAdvertiseAddress(hc) which then call net.ParseIP( and ignores the error case.

Can we let findAdvertiseAddress to use netip.ParseAddr( and return an error when the input is not parseable?

---

Addressed